src/plugins/events/events.ts

```diff
@@ -69,14 +74,41 @@ export class Events {
     return this.dispatch.length > 0;
   }
 
+  updateTurnContext(ctx: Ctx) {
+    this.currentPhase = ctx.phase;
+    this.currentTurn = ctx.turn;
+  }
+
   /**
    * Updates ctx with the triggered events.
    * @param {object} state - The state object { G, ctx }.
    */
   update(state: State) {
     for (let i = 0; i < this.dispatch.length; i++) {
+      const endedTurns = this.currentTurn - this.initialTurn;
+      // This is an arbitrarily large number.
+      const threshold = state.ctx.numPlayers * 10;
+
+      // This protects against potential infinite loops if specific events are called on hooks.
+      // The moment we exceed the defined threshold, we just bail out of all phases.
+      if (endedTurns > threshold) {
+        state = { ...state, ctx: { ...state.ctx, phase: null } };
+        break;
```

Comment:
It seems like a good idea. I didn't do it because it is not currently being done in other similar situations. It would be necessary to define how to log it.

Reply:
We have a little logging utility that you can import. Here it is in `core/flow.ts`:

https://github.com/boardgameio/boardgame.io/blob/bfc8a1f2701fd14627c99f38122abc294aa7c16e/src/core/flow.ts#L20

https://github.com/boardgameio/boardgame.io/blob/bfc8a1f2701fd14627c99f38122abc294aa7c16e/src/core/flow.ts#L68

https://github.com/boardgameio/boardgame.io/blob/bfc8a1f2701fd14627c99f38122abc294aa7c16e/src/core/flow.ts#L498-L500

Reply:
Logging should be info or error?

Reply:
Good question. I think in my head it’s actually a warning which we don’t have set-up but let’s go with `info` as the action doesn’t actually error currently.